diff --git a/folly/experimental/exception_tracer/CMakeLists.txt b/folly/experimental/exception_tracer/CMakeLists.txt
index 6156f53620..19929e3b98 100644
--- a/folly/experimental/exception_tracer/CMakeLists.txt
+++ b/folly/experimental/exception_tracer/CMakeLists.txt
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-if (FOLLY_HAVE_ELF AND FOLLY_HAVE_DWARF)
+if (NOT FOLLY_NO_EXCEPTION_TRACER AND FOLLY_HAVE_ELF AND FOLLY_HAVE_DWARF)
   add_library(
     folly_exception_tracer_base
     ExceptionTracer.cpp
