diff --git a/folly/system/AtFork.cpp b/folly/system/AtFork.cpp
index e888e52..27f37e7 100644
--- a/folly/system/AtFork.cpp
+++ b/folly/system/AtFork.cpp
@@ -16,6 +16,9 @@

 #include <folly/system/AtFork.h>

+#include <stdexcept>
+#include <system_error>
+
 #include <folly/ScopeGuard.h>
 #include <folly/lang/Exception.h>
 #include <folly/portability/PThread.h>
diff --git a/CMakeLists.txt b/CMakeLists.txt
index ddfdc76..f180197 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -174,6 +174,7 @@ REMOVE_MATCHES_FROM_LISTS(files hfiles

 # Exclude exception tracer, which is necessary to statically link libstdc++
 if (${FOLLY_NO_EXCEPTION_TRACER})
+    message(WARNING "Exclude exception tracer, which is necessary to statically link libstdc++")
   REMOVE_MATCHES_FROM_LISTS(files hfiles
     MATCHES
       "^${FOLLY_DIR}/experimental/exception_tracer/"
diff --git a/folly/experimental/exception_tracer/CMakeLists.txt b/folly/experimental/exception_tracer/CMakeLists.txt
index 6156f53..19929e3 100644
--- a/folly/experimental/exception_tracer/CMakeLists.txt
+++ b/folly/experimental/exception_tracer/CMakeLists.txt
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.

-if (FOLLY_HAVE_ELF AND FOLLY_HAVE_DWARF)
+if (NOT FOLLY_NO_EXCEPTION_TRACER AND FOLLY_HAVE_ELF AND FOLLY_HAVE_DWARF)
   add_library(
     folly_exception_tracer_base
     ExceptionTracer.cpp
