diff --git a/folly/experimental/symbolizer/Elf.cpp b/folly/experimental/symbolizer/Elf.cpp
index 903712e5e..b2c10f17d 100644
--- a/folly/experimental/symbolizer/Elf.cpp
+++ b/folly/experimental/symbolizer/Elf.cpp
@@ -30,6 +30,7 @@
 #include <folly/lang/CString.h>
 #include <folly/portability/Config.h>
 #include <folly/portability/SysMman.h>
+#include <folly/portability/Unistd.h>
 
 #if FOLLY_HAVE_ELF
 
