diff --git a/cpp/src/arrow/csv/parser.cc b/cpp/src/arrow/csv/parser.cc
index da3472a9d..690e94119 100644
--- a/cpp/src/arrow/csv/parser.cc
+++ b/cpp/src/arrow/csv/parser.cc
@@ -192,6 +192,8 @@ class BlockParserImpl {
         max_num_rows_(max_num_rows),
         batch_(num_cols) {}
 
+  const ParseOptions& getParseOptions() const { return options_; }
+
   const DataBatch& parsed_batch() const { return batch_; }
 
   int64_t first_row_num() const { return first_row_; }
@@ -698,5 +700,9 @@ int32_t SkipRows(const uint8_t* data, uint32_t size, int32_t num_rows,
   return skipped_rows;
 }
 
+const ParseOptions& BlockParser::getParseOptions() const {
+  return impl_->getParseOptions();
+}
+
 }  // namespace csv
 }  // namespace arrow
