diff --git a/cpp/src/arrow/csv/parser.h b/cpp/src/arrow/csv/parser.h
index c73e52ce8..fdbb23f90 100644
--- a/cpp/src/arrow/csv/parser.h
+++ b/cpp/src/arrow/csv/parser.h
@@ -218,6 +218,8 @@ class ARROW_EXPORT BlockParser {
     return parsed_batch().VisitLastRow(std::forward<Visitor>(visit));
   }
 
+  const ParseOptions& getParseOptions() const;
+
  protected:
   std::unique_ptr<BlockParserImpl> impl_;
 
