diff --git a/cpp/src/arrow/csv/options.h b/cpp/src/arrow/csv/options.h
index 7723dcedc..b62cf88f0 100644
--- a/cpp/src/arrow/csv/options.h
+++ b/cpp/src/arrow/csv/options.h
@@ -44,6 +44,10 @@ struct ARROW_EXPORT ParseOptions {
 
   /// Field delimiter
   char delimiter = ',';
+  /// Collection delimiter
+  char collection_delimiter = ';';
+  /// Map key delimiter
+  char map_key_delimiter = ':';
   /// Whether quoting is used
   bool quoting = true;
   /// Quoting character (if `quoting` is true)
