diff --git a/cpp/src/arrow/csv/converter.h b/cpp/src/arrow/csv/converter.h
index 639f692f2..04832d9d0 100644
--- a/cpp/src/arrow/csv/converter.h
+++ b/cpp/src/arrow/csv/converter.h
@@ -78,5 +78,40 @@ class ARROW_EXPORT DictionaryConverter : public Converter {
   std::shared_ptr<DataType> value_type_;
 };
 
+class ARROW_EXPORT ListConverter : public Converter {
+ public:
+  ListConverter(const std::shared_ptr<DataType>& value_type,
+                      const ConvertOptions& options, MemoryPool* pool);
+
+  // If the list length goes above this value, conversion will fail
+  // with Status::IndexError.
+  virtual void SetMaxCardinality(int32_t max_length) = 0;
+
+  static Result<std::shared_ptr<ListConverter>> Make(
+      const std::shared_ptr<DataType>& value_type, const ConvertOptions& options,
+      MemoryPool* pool = default_memory_pool());
+
+ protected:
+  std::shared_ptr<DataType> value_type_;
+};
+
+class ARROW_EXPORT MapConverter : public Converter {
+ public:
+  MapConverter(const std::shared_ptr<DataType>& key_type, const std::shared_ptr<DataType>& item_type_,
+                      const ConvertOptions& options, MemoryPool* pool);
+
+  // If the map length goes above this value, conversion will fail
+  // with Status::IndexError.
+  virtual void SetMaxCardinality(int32_t max_length) = 0;
+
+  static Result<std::shared_ptr<MapConverter>> Make(
+      const std::shared_ptr<DataType>& key_type, const std::shared_ptr<DataType>& item_type_, const ConvertOptions& options,
+      MemoryPool* pool = default_memory_pool());
+
+ protected:
+  std::shared_ptr<DataType> key_type_;
+  std::shared_ptr<DataType> item_type_;
+};
+
 }  // namespace csv
 }  // namespace arrow
